### ISSUE TITLE
[speech-to-text] Account for frankfurt endpoint for wss connection

### ISSUE THREAD

**Issue by miltatibm** [watson-developer-cloud/unity-sdk#353]
Customer Issue
Migrated tts service from US-South to Germany eu-gb
Applied the correct uri and correct credentials
Getting the following error

> [03/27/2018 11:00:05][WSConnector][ERROR] Caught WebSocket exception: System.ArgumentException: The scheme part is not 'ws' or 'wss'.
Parameter name: url
  at WebSocketSharp.WebSocket..ctor (System.String url, System.String[] protocols) [0x00000] in <filename unknown>:0 
  at IBM.Watson.DeveloperCloud.Connection.WSConnector.SendMessages () [0x00010] in /Users/xxxxxxx/Assets/Watson/Scripts/Connection/WSConnector.cs:309 
UnityEngine.Debug:LogError(Object)
IBM.Watson.DeveloperCloud.Debug.DebugReactor:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Debug/DebugReactor.cs:60)
IBM.Watson.DeveloperCloud.Logging.LogSystem:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Logging/Logger.cs:196)
IBM.Watson.DeveloperCloud.Logging.Log:Error(String, String, Object[]) (at Assets/Watson/Scripts/Logging/Logger.cs:261)
IBM.Watson.DeveloperCloud.Connection.WSConnector:SendMessages() (at Assets/Watson/Scripts/Connection/WSConnector.cs:352) 

>

**Comment by mediumTaj** [comment]
This has to do with the TLS 1.2 switch. I am still investigating this. We had tested the TLS 1.2 endpoint with another client and it worked as expected. When we disabled TLS 1.0, wss connections failed in the Unity SDK. Currently I am converting the http link to a wss link and temporarily pointing it to the TLS 1.0 endpoint until I could investigate. Other regions are not accounted for though. I asked the platform team to enable TLS 1.0 endpoint for the fra region. Once we have that I can revise the code to convert the fra endpoint to point to TLS 1.0.

**Comment by miltatibm** [comment]
@mediumTaj Yup he just tried the wss url with and w/out the tls1.0 and it just disconnects

> 03/28/2018 20:09:09][SpeechToText.OnListenClosed()][DEBUG] OnListenClosed(), State = DISCONNECTED
UnityEngine.Debug:Log(Object)
IBM.Watson.DeveloperCloud.Debug.DebugReactor:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Debug/DebugReactor.cs:68)
IBM.Watson.DeveloperCloud.Logging.LogSystem:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Logging/Logger.cs:196)
IBM.Watson.DeveloperCloud.Logging.Log:Debug(String, String, Object[]) (at Assets/Watson/Scripts/Logging/Logger.cs:216)
IBM.Watson.DeveloperCloud.Services.SpeechToText.v1.SpeechToText:OnListenClosed(WSConnector) (at Assets/Watson/Scripts/Services/SpeechToText/v1/SpeechToText.cs:810)
IBM.Watson.DeveloperCloud.Connection.<ProcessReceiveQueue>c__Iterator0:MoveNext() (at Assets/Watson/Scripts/Connection/WSConnector.cs:296)
IBM.Watson.DeveloperCloud.Utilities.Routine:MoveNext() (at Assets/Watson/Scripts/Utilities/Runnable.cs:128)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)>

**Comment by miltatibm** [comment]
@mediumTaj yes they just tried with and without the endpoint and it just disconnects

> 03/28/2018 20:09:09][SpeechToText.OnListenClosed()][DEBUG] OnListenClosed(), State = DISCONNECTED
UnityEngine.Debug:Log(Object)
IBM.Watson.DeveloperCloud.Debug.DebugReactor:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Debug/DebugReactor.cs:68)
IBM.Watson.DeveloperCloud.Logging.LogSystem:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Logging/Logger.cs:196)
IBM.Watson.DeveloperCloud.Logging.Log:Debug(String, String, Object[]) (at Assets/Watson/Scripts/Logging/Logger.cs:216)
IBM.Watson.DeveloperCloud.Services.SpeechToText.v1.SpeechToText:OnListenClosed(WSConnector) (at Assets/Watson/Scripts/Services/SpeechToText/v1/SpeechToText.cs:810)
IBM.Watson.DeveloperCloud.Connection.<ProcessReceiveQueue>c__Iterator0:MoveNext() (at Assets/Watson/Scripts/Connection/WSConnector.cs:296)
IBM.Watson.DeveloperCloud.Utilities.Routine:MoveNext() (at Assets/Watson/Scripts/Utilities/Runnable.cs:128)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr) <

**Comment by miltatibm** [comment]
Update from the customer

I tried both wss options, with the tls10 and without. Neither worked. I don't get an error anymore, the service just immediately disconnects. 03/28/2018 20:09:09][SpeechToText.OnListenClosed()][DEBUG] OnListenClosed(), State = DISCONNECTED
UnityEngine.Debug:Log(Object)
IBM.Watson.DeveloperCloud.Debug.DebugReactor:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Debug/DebugReactor.cs:68)
IBM.Watson.DeveloperCloud.Logging.LogSystem:ProcessLog(LogRecord) (at Assets/Watson/Scripts/Logging/Logger.cs:196)
IBM.Watson.DeveloperCloud.Logging.Log:Debug(String, String, Object[]) (at Assets/Watson/Scripts/Logging/Logger.cs:216)
IBM.Watson.DeveloperCloud.Services.SpeechToText.v1.SpeechToText:OnListenClosed(WSConnector) (at Assets/Watson/Scripts/Services/SpeechToText/v1/SpeechToText.cs:810)
IBM.Watson.DeveloperCloud.Connection.<ProcessReceiveQueue>c__Iterator0:MoveNext() (at Assets/Watson/Scripts/Connection/WSConnector.cs:296)
IBM.Watson.DeveloperCloud.Utilities.Routine:MoveNext() (at Assets/Watson/Scripts/Utilities/Runnable.cs:128)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

**Comment by mediumTaj** [comment]
@miltatibm The platform team has assessed the work that needs to be done to enable a TLS 1.0 endpoint in the Frankfurt region and we've decided that it's too much work to put into a temporary patch. I'm trying to move forward with the Unity team to try to find a permanent fix for this. Until then, is there any way the client can switch back to the Dallas region?

**Comment by miltatibm** [comment]
@mediumTaj I let the customer know about the complexity and suggested to try using US-South as an option. No response yet. Thanks!

**Comment by mediumTaj** [comment]
@miltatibm I worked with some engineers at Unity on the TLS issue. It will be fixed in `2018.2`. You can try the beta out here
https://unity3d.com/unity/beta-download

There will need to be some revisions to the sdk though.  
WSConnector.cs line 319, add

```
ws.SslConfiguration.EnabledSslProtocols = SslProtocols.Tls12;
```

also in the `FixupURL` method in WSConnector.cs add
```
else if (URL.StartsWith(“http://stream-fra.”)
    URL = URL.Replace("http://stream-fra.”, “ws://stream-fra.”);
else if (URL.StartsWith(“https://stream-fra.”)
    URL = URL.Replace("https://stream-fra.”, “wss://stream-fra.”);
```

I'll have these changes in another PR soon hopefully. Please let me know if they are able to test this

**Comment by mediumTaj** [comment]
This has been addressed in https://github.com/watson-developer-cloud/unity-sdk/releases/tag/2.4.1

Unity 2018.2 supports TLS 1.2 so we can use websockets across different regions.
